Incorporate calibration optimization functions

padilla410 opened the pull request:
The goal of this PR is to get a working set of calibration functions that will allow me to optimize the calibration of the Missouri reservoir models. There are a lot of commits in here, but the important ones are from today (2022-06-22). I would like you to complete the following:
* Review the functions in `3_calibrate/src/calibration_utils.R`
* Run `test_sim/test_simulation_run.R` and verify that the functions work
* Comment on which approach under "Next Steps" would be appropriate
* Answer my question under general questions

Along with the committed changes I included a subfolder called `test_sim`. This subfolder contains all of the data and files needed to run calibrations for two reservoirs: Pomme De Terre and Lake Stockton. It also includes `Rmd` files to evaluate the calibration of each run after you run the optimization (if you're interested). To run this locally you will need to open `test_sim/test_simulation_run.R` and run the code. It takes around 27 min to run the optimization for Pomme De Terre. I did not time the optimization for Lake Stockton.

Known issues:
* Duplicate `run_glm` functions: one in  `2_run/src/run_glm3.R` and another in `3_calibrate/src/calibration_utils.R`.
  * It looks to me like folks often write repo-specific `run_glm` functions rather than using `glmtools::run_glm` (see [`process-models`](https://github.com/USGS-R/lake-temperature-process-models/blob/8cae44f02fbacb9aa1851a1439f4a208e8304a3a/2_run/src/run_glm3.R#L75-L212) and [`process-models-old`](https://github.com/USGS-R/lake-temperature-process-models-old/blob/79ccf6f2c7a8631549e8c935a26133db9ab4f968/3_run/src/run_glm_utils.R#L114-L131). Focusing on what exactly this repo's `run_glm` should return seemed like it should be a different PR.
  * For the test optimization, I used a slightly modified version of `run_glm` from `process-models-old` (updates from `GLM` to `GLM3`)
* Messy file structure - this will be resolved when the pipeline is formalized

Next steps:
* I think we have two options if this PR looks good:
  * Develop the calibration portion of the pipeline
  * Focus on what this repo-specific `run_glm` function should look like

General questions:
* What exactly is the `cal_parscale` argument in `run_glm_cal` about? It gets passed to the `control` argument in `optim`. I would like to know a bit more so I can improve the function documentation.

jread-usgs reviewed (commented):
Yes, this approach will work for calibration. I had a few notes in here. 

There are a lot of commits/files that are unrelated to the ask of the PR review. I didn't review them. 

as you note, `parscale` is an input to the base/`stats` `optim()` function, which we use `"Nelder-Mead"` method. From the docs:

> parscale
A vector of scaling values for the parameters. Optimization is performed on par/parscale and these should be comparable in the sense that a unit change in any element produces about a unit change in the scaled value. Not used

padilla410 commented on README.md:
```diff
@@ -8,7 +8,7 @@ GLM process model pipeline for 8 priority reservoirs in Missouri. This pipeline
 * NLDAS driver files (stored on Caldera)
   * _e.g._, `'1_prep/in/NLDAS_time[0.379366]_x[231]_y[167].csv'`
```
Updated

padilla410 commented on test_sim/test_simulation_run.R:
```diff
@@ -0,0 +1,27 @@
+
+# prep calibration parameters
+cal_params <- c('cd','sw_factor','coef_mix_hyp')
+cal_parscale <- c('cd' = 0.0001, 'sw_factor' = 0.02, 'coef_mix_hyp' = 0.01)
```
I went with the latter

padilla410 commented on test_sim/test_simulation_run.R:
```diff
@@ -0,0 +1,27 @@
+caldata_fl <- file.path(sim_dir, 'field_data_nhdhr_106716325.rds') #'1_prep/out/field_data_nhdhr_106716325.rds'
+
+# run optimization for Pomme De Terre Lake
+run_glm_cal(nml_file, sim_dir, cal_params, cal_parscale, caldata_fl)
```
oops :)

I'll fix that

padilla410 commented on test_sim/test_simulation_run.R:
```diff
@@ -0,0 +1,27 @@
+source('3_calibrate/src/calibration_utils.R')
+
+# prep calibration parameters
+cal_params <- c('cd','sw_factor','coef_mix_hyp')
```
The addition of `Kw` makes sense, but I was surprised to see the removal of `coef_hyp_mixing`. Would you mind adding a little more context for this swap?

padilla410 commented on 3_calibrate/src/calibration_utils.R:
```diff
@@ -0,0 +1,101 @@
+  nlm_obj <- glmtools::set_nml(nml_obj, arg_list = setNames(as.list(out$par), cal_params))
+
+  # write the rmse and other details into a new block in the nml "results"
+  nml_obj$results <- list(rmse = out$value,
```
Added

padilla410 commented:
I have incorporated all suggested edits except for the swapping of `Kw` for `coef_mix_hyp` in the optimization. I did not resolve that conversation because I have lingering questions and I think it will be an easy fix in a future PR. I moved that conversation to issue #3.
